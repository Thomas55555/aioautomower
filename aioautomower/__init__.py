"""Automower library using aiohttp."""
from .rest import *
from .session import AutomowerSession
