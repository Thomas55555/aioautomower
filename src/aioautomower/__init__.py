"""Automower library using aiohttp."""

__all__ = [
    "auth",
    "const",
    "exceptions",
    "model",
    "session",
    "utils",
]
