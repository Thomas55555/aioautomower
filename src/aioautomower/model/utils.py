"""Helper utils for Automower Connect API."""

from datetime import UTC, datetime

from aioautomower import tz_util


def convert_timestamp_to_aware_datetime(timestamp: int) -> datetime | None:
    """Convert the timestamp to an aware datetime object.

    The Python datetime library expects timestamps to be anchored in UTC,
    however, the automower timestamps are anchored in local time. So we convert
    the timestamp to a datetime and replace the timezone with the local time.
    After that we convert the timezone to UTC.
    """
    if timestamp == 0:
        return None
    if timestamp > 32503680000:
        # This will break on January 1th 3000. If mankind still exists there
        # please fix it.
        return datetime.fromtimestamp(timestamp / 1000, tz=UTC).replace(
            tzinfo=tz_util.MOWER_TIME_ZONE
        )
    return datetime.fromtimestamp(timestamp, tz=UTC).replace(
        tzinfo=tz_util.MOWER_TIME_ZONE
    )
