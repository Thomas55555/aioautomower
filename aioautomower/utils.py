"""Utils for Husqvarna Automower."""
import logging
import time
from urllib.parse import quote_plus, urlencode

import aiohttp
import jwt

from .const import AUTH_API_REVOKE_URL, AUTH_API_TOKEN_URL, AUTH_HEADERS
from .model import JWT
from .rest import TokenError

_LOGGER = logging.getLogger(__name__)


async def async_structure_token(access_token) -> JWT:
    """Decode JWT and convert to dataclass."""
    token_decoded = jwt.decode(access_token, options={"verify_signature": False})
    return JWT(**token_decoded)


async def async_get_access_token(client_id, client_secret) -> dict:
    """Get an access token from the Authentication API with client credentials.

    This grant type is intended only for you. If you want other
    users to use your application, then they should login using Authorization
    Code Grant.
    """
    auth_data = urlencode(
        {
            "grant_type": "client_credentials",
            "client_id": client_id,
            "client_secret": client_secret,
        },
        quote_via=quote_plus,
    )
    async with aiohttp.ClientSession(headers=AUTH_HEADERS) as session, session.post(
        AUTH_API_TOKEN_URL, data=auth_data
    ) as resp:
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status get access token: %s", result)
        if resp.status == 200:
            result = await resp.json(encoding="UTF-8")
            result["expires_at"] = result["expires_in"] + time.time()
        if resp.status >= 400:
            raise TokenError(
                f"""The token is invalid, respone from
                    Husqvarna Automower API: {result}"""
            )
    result["status"] = resp.status
    return result


async def async_invalidate_access_token(
    valid_access_token, access_token_to_invalidate
) -> dict:
    """Invalidate the token.

    :param str valid_access_token: A working access token to authorize this request.
    :param str access_token_to_delete: An access token to invalidate,
    can be th same like the first argument.
    """
    headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": f"Bearer {valid_access_token}",
        "Accept": "*/*",
    }
    async with aiohttp.ClientSession(headers=headers) as session, session.post(
        AUTH_API_REVOKE_URL, data=(f"token={access_token_to_invalidate}")
    ) as resp:
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status delete token: %s", resp.status)
        if resp.status >= 400:
            resp.raise_for_status()
            _LOGGER.error("Response body delete token: %s", result)
    return result
