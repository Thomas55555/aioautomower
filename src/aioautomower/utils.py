"""Utils for Husqvarna Automower."""

import logging
import time
from typing import Any, Mapping, cast
from urllib.parse import quote_plus, urlencode

import aiohttp
import jwt

from .const import AUTH_API_REVOKE_URL, AUTH_API_TOKEN_URL, AUTH_HEADERS, ERRORCODES
from .exceptions import ApiException
from .model import JWT, MowerAttributes, MowerList, snake_case

_LOGGER = logging.getLogger(__name__)


def structure_token(access_token: str) -> JWT:
    """Decode JWT and convert to dataclass."""
    token_decoded = jwt.decode(access_token, options={"verify_signature": False})
    return JWT.from_dict(token_decoded)


async def async_get_access_token(client_id: str, client_secret: str) -> dict[str, str]:
    """Get an access token from the Authentication API with client credentials.

    This grant type is intended only for you. If you want other
    users to use your application, then they should login using Authorization
    Code Grant.
    """
    auth_data = urlencode(
        {
            "grant_type": "client_credentials",
            "client_id": client_id,
            "client_secret": client_secret,
        },
        quote_via=quote_plus,
    )
    async with (
        aiohttp.ClientSession(headers=AUTH_HEADERS) as session,
        session.post(AUTH_API_TOKEN_URL, data=auth_data) as resp,
    ):
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status get access token: %s", result)
        if resp.status == 200:
            result = await resp.json(encoding="UTF-8")
            result["expires_at"] = result["expires_in"] + time.time()
        if resp.status >= 400:
            raise ApiException(
                f"""The token is invalid, response from
                    Husqvarna Automower API: {result}"""
            )
    result["status"] = resp.status
    return cast(dict[str, str], result)


async def async_invalidate_access_token(
    valid_access_token: str, access_token_to_invalidate: str
) -> dict[str, str]:
    """Invalidate the token.

    :param str valid_access_token: A working access token to authorize this request.
    :param str access_token_to_delete: An access token to invalidate,
    can be th same like the first argument.
    """
    headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": f"Bearer {valid_access_token}",
        "Accept": "*/*",
    }
    async with (
        aiohttp.ClientSession(headers=headers) as session,
        session.post(
            AUTH_API_REVOKE_URL, data=(f"token={access_token_to_invalidate}")
        ) as resp,
    ):
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status delete token: %s", resp.status)
        if resp.status >= 400:
            resp.raise_for_status()
            _LOGGER.error("Response body delete token: %s", result)
    return cast(dict[str, str], result)


def mower_list_to_dictionary_dataclass(
    mower_list: Mapping[Any, Any],
) -> dict[str, MowerAttributes]:
    """Convert mower data to a dictionary DataClass."""
    mowers_list = MowerList.from_dict(mower_list)
    mowers_dict = {}
    for mower in mowers_list.data:
        mowers_dict[mower.id] = mower.attributes
    return mowers_dict


def error_key_list() -> list[str]:
    """Create a list with all possible error keys."""
    codes = [snake_case(error_text) for error_text in ERRORCODES.values()]
    return sorted(codes)


def error_key_dict() -> dict[str, str]:
    """Create a dictionary with error keys and a human friendly text."""
    codes = {}
    for error_text in ERRORCODES.values():
        codes[snake_case(error_text)] = error_text
    return codes
