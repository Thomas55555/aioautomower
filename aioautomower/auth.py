"""Module for AbstractAuth for Husqvarna Automower."""

from abc import ABC, abstractmethod
from collections.abc import Mapping
from http import HTTPStatus
import logging
from typing import Any, Optional

import aiohttp
from aiohttp import ClientError, ClientResponse, ClientResponseError, ClientSession

from .const import API_BASE_URL, AUTH_HEADER_FMT, EVENT_TYPES, WS_URL
from .exceptions import ApiException, ApiForbiddenException, AuthException
from .utils import async_structure_token

ERROR = "error"
STATUS = "status"
MESSAGE = "message"

_LOGGER = logging.getLogger(__name__)


class AbstractAuth(ABC):
    """Abstract class to make authenticated requests."""

    def __init__(self, websession: ClientSession, host: str) -> None:
        """Initialize the auth."""
        self._websession = websession
        self._host = host if host is not None else API_BASE_URL
        self._client_id = None
        self.loop = websession.loop
        self.ws_update_cbs = []
        self.ws_data = {}

    @abstractmethod
    async def async_get_access_token(self) -> str:
        """Return a valid access token."""

    async def request(
        self, method: str, url: str, **kwargs: Optional[Mapping[str, Any]]
    ) -> ClientResponse:
        """Make a request."""
        headers = await self.headers()
        if not (url.startswith("http://") or url.startswith("https://")):
            url = f"{self._host}/{url}"
        _LOGGER.debug("request[%s]=%s %s", method, url, kwargs.get("params"))
        if method != "get" and "json" in kwargs:
            _LOGGER.debug("request[post json]=%s", kwargs["json"])
        return await self._websession.request(method, url, **kwargs, headers=headers)

    async def get(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a get request."""
        try:
            resp = await self.request("get", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def get_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a get request and return json response."""
        resp = await self.get(url, **kwargs)
        try:
            result = await resp.json(encoding="UTF-8")
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server return malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def post(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a post request."""
        try:
            resp = await self.request("post", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def post_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a post request and return a json response."""
        resp = await self.post(url, **kwargs)
        try:
            result = await resp.json()
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server returned malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def patch(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a post request."""
        try:
            resp = await self.request("patch", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def patch_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a post request and return a json response."""
        resp = await self.patch(url, **kwargs)
        try:
            result = await resp.json()
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server returned malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def _async_get_access_token(self) -> None:
        """Requests a new access token."""
        try:
            return await self.async_get_access_token()
        except ClientError as err:
            raise AuthException(f"Access token failure: {err}") from err

    async def headers(self) -> None:
        """Generate headers for ReST requests."""
        access_token = await self._async_get_access_token()
        if not self._client_id:
            token_structured = await async_structure_token(access_token)
            self._client_id = token_structured.client_id
        return {
            "Authorization": f"Bearer {access_token}",
            "Authorization-Provider": "husqvarna",
            "Content-Type": "application/vnd.api+json",
            "X-Api-Key": self._client_id,
        }

    @staticmethod
    async def _raise_for_status(resp: ClientResponse) -> ClientResponse:
        """Raise exceptions on failure methods."""
        detail = await AbstractAuth._error_detail(resp)
        try:
            resp.raise_for_status()
        except ClientResponseError as err:
            if err.status == HTTPStatus.FORBIDDEN:
                raise ApiForbiddenException(
                    f"Forbidden response from API: {err}"
                ) from err
            if err.status == HTTPStatus.UNAUTHORIZED:
                raise AuthException(f"Unable to authenticate with API: {err}") from err
            detail.append(err.message)
            raise ApiException(": ".join(detail)) from err
        except ClientError as err:
            raise ApiException(f"Error from API: {err}") from err
        return resp

    @staticmethod
    async def _error_detail(resp: ClientResponse) -> list[str]:
        """Return an error message string from the API response."""
        if resp.status < 400:
            return []
        try:
            result = await resp.json()
            error = result.get(ERROR, {})
        except ClientError:
            return []
        message = ["Error from API", f"{resp.status}"]
        if STATUS in error:
            message.append(f"{error[STATUS]}")
        if MESSAGE in error:
            message.append(error[MESSAGE])
        return message

    def register_ws_callback(self, callback):
        """Register a websocket update callback.

        :param func callback: Callback fired on data updates. Takes one dict argument which is the latest websocket data.
        """
        if callback not in self.ws_update_cbs:
            self.ws_update_cbs.append(callback)

    def _schedule_ws_callback(self, cb, delay=0.0):
        """Schedule websocket callback."""
        self.loop.call_later(delay, cb, self.ws_data)

    def _schedule_ws_callbacks(self):
        """Schedule all websocket callbacks."""
        for cb in self.ws_update_cbs:
            self._schedule_ws_callback(cb)

    async def websocket(self) -> dict[str, dict]:
        """Start a websocket conenction."""
        async with self._websession.ws_connect(
            url=WS_URL,
            headers={
                "Authorization": AUTH_HEADER_FMT.format(
                    await self._async_get_access_token()
                )
            },
            heartbeat=60,
        ) as ws:
            async for msg in ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    msg_dict = msg.json()
                    if "type" in msg_dict:
                        if msg_dict["type"] in EVENT_TYPES:
                            _LOGGER.debug(
                                "Got %s, data: %s", msg_dict["type"], msg_dict
                            )
                            self.ws_data = msg_dict
                            self._schedule_ws_callbacks()
                        else:
                            _LOGGER.warning(
                                "Received unknown ws type %s", msg_dict["type"]
                            )
                    elif "ready" in msg_dict and "connectionId" in msg_dict:
                        _LOGGER.debug(
                            "Websocket ready=%s (id='%s')",
                            msg_dict["ready"],
                            msg_dict["connectionId"],
                        )
                    else:
                        _LOGGER.debug("Discarded websocket response: %s", msg_dict)
                elif msg.type == aiohttp.WSMsgType.ERROR:
                    _LOGGER.debug("Received ERROR")
                    break
                elif msg.type == aiohttp.WSMsgType.CONTINUATION:
                    _LOGGER.debug("Received CONTINUATION")
                elif msg.type == aiohttp.WSMsgType.BINARY:
                    _LOGGER.debug("Received BINARY")
                elif msg.type == aiohttp.WSMsgType.PING:
                    _LOGGER.debug("Received PING")
                elif msg.type == aiohttp.WSMsgType.PONG:
                    _LOGGER.debug("Received PONG")
                elif msg.type == aiohttp.WSMsgType.CLOSE:
                    _LOGGER.debug("Received CLOSE")
                    await ws.close()
                elif msg.type == aiohttp.WSMsgType.CLOSING:
                    _LOGGER.debug("Received CLOSING")
                elif msg.type == aiohttp.WSMsgType.CLOSED:
                    _LOGGER.debug("Received CLOSED")
                else:
                    _LOGGER.debug("Received msg.type=%d", msg.type)
