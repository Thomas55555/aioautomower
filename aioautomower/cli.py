#!/usr/bin/env python3
import argparse
import asyncio
import logging
import signal
import time

from aiohttp import ClientSession

from aioautomower.auth import AbstractAuth
from aioautomower.session import AutomowerSession
import aioautomower.utils

from .const import API_BASE_URL

_LOGGER = logging.getLogger(__name__)


async def run_tester(client_id: str, client_secret: str):
    automower_api = AutomowerSession(
        AsyncConfigEntryAuth(ClientSession(), client_id, client_secret), poll=True
    )
    automower_api.register_data_callback(
        lambda x: logging.info("data callback;%s", x),
    )

    await automower_api.connect()

    def sigusr1():
        asyncio.ensure_future(automower_api.invalidate_token())

    def sigusr2():
        asyncio.ensure_future(automower_api.get_status())

    loop = asyncio.get_event_loop()
    loop.add_signal_handler(signal.SIGUSR1, sigusr1)
    loop.add_signal_handler(signal.SIGUSR2, sigusr2)

    while True:
        await asyncio.sleep(0.1)


def main():
    """Tester for the Husqvarna Automower API.

    The tester will login using client_id and client_secret and connect to a
    websocket listening for mower updates.

    The tester listens to two signals on which it performs the following
    requests:
    SIGUSR1: Invalidate token
    SIGUSR2: Get status
    """

    parser = argparse.ArgumentParser(
        description=main.__doc__, formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument(
        "-s", "--client_id", required=True, help="Husqvarna Application key"
    )
    parser.add_argument(
        "-k", "--client_secret", required=True, help="Husqvarna Application secret"
    )

    args = parser.parse_args()

    logging.basicConfig(level="DEBUG", format="%(asctime)s;%(levelname)s;%(message)s")
    asyncio.run(run_tester(args.client_id, args.client_secret))


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide Husqvarna Automower authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        client_id: str,
        client_secret: str,
    ) -> None:
        """Initialize Husqvarna Automower auth."""
        super().__init__(websession, API_BASE_URL)
        self.client_id = client_id
        self.client_secret = client_secret
        self.token = None
        self.websession = websession

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self.token:
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
            token_structured = await aioautomower.utils.async_structure_token(
                self.token["access_token"]
            )
            _LOGGER.debug("token_structured.exp: %s", token_structured.exp)
            _LOGGER.debug(time.time())
        if token_structured.exp < time.time():
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
        return self.token["access_token"]
