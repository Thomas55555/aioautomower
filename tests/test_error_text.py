"""Tests for asynchronous Python client for aioautomower."""

import json


from aioautomower.utils import (
    mower_list_to_dictionary_dataclass,
    error_key_list,
    error_key_dict,
)
from tests import load_fixture


MOWER_ID = "c7233734-b219-4287-a173-08e3643f89f0"


async def test_error_key() -> None:
    """Test translating an error code to an error key."""
    mower_fixture = load_fixture("high_feature_mower.json")
    mower_python = json.loads(mower_fixture)
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert mowers[MOWER_ID].mower.error_key is None

    mower_python["data"][0]["attributes"]["mower"]["errorCode"] = 1
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert mowers[MOWER_ID].mower.error_key == "outside_working_area"

    mower_python["data"][0]["attributes"]["mower"]["errorCode"] = 8
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert mowers[MOWER_ID].mower.error_key == "wrong_pin_code"

    mower_python["data"][0]["attributes"]["mower"]["errorCode"] = 18
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert mowers[MOWER_ID].mower.error_key == "collision_sensor_problem_rear"

    mower_python["data"][0]["attributes"]["mower"]["errorCode"] = 78
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert (
        mowers[MOWER_ID].mower.error_key
        == "slipped_mower_has_slipped_situation_not_solved_with_moving_pattern"
    )


async def test_error_keys_snapshot(snapshot) -> None:
    """Make a snapshot of the error keys."""
    assert error_key_list() == snapshot


async def test_error_key_dict_snapshot(snapshot) -> None:
    """Make a snapshot of the error key dictionary."""
    assert error_key_dict() == snapshot
