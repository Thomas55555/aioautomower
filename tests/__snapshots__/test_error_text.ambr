# serializer version: 1
# name: test_error_key_dict_snapshot
  dict({
    'alarm_mower_in_motion': 'Alarm! Mower in motion',
    'alarm_mower_lifted': 'Alarm! Mower lifted',
    'alarm_mower_stopped': 'Alarm! Mower stopped',
    'alarm_mower_switched_off': 'Alarm! Mower switched off',
    'alarm_mower_tilted': 'Alarm! Mower tilted',
    'alarm_outside_geofence': 'Alarm! Outside geofence',
    'angular_sensor_problem': 'Angular sensor problem',
    'battery_problem': 'Battery problem',
    'battery_restriction_due_to_ambient_temperature': 'Battery restriction due to ambient temperature',
    'can_error': 'CAN error',
    'charging_current_too_high': 'Charging current too high',
    'charging_station_blocked': 'Charging station blocked',
    'charging_system_problem': 'Charging system problem',
    'collision_sensor_defect': 'Collision sensor defect',
    'collision_sensor_error': 'Collision sensor error',
    'collision_sensor_problem_front': 'Collision sensor problem, front',
    'collision_sensor_problem_rear': 'Collision sensor problem, rear',
    'com_board_not_available': 'Com board not available',
    'communication_circuit_board_sw_must_be_updated': 'Communication circuit board SW must be updated',
    'complex_working_area': 'Complex working area',
    'connection_changed': 'Connection changed',
    'connection_not_changed': 'Connection NOT changed',
    'connectivity_problem': 'Connectivity problem',
    'connectivity_settings_restored': 'Connectivity settings restored',
    'cutting_drive_motor_1_defect': 'Cutting drive motor 1 defect',
    'cutting_drive_motor_2_defect': 'Cutting drive motor 2 defect',
    'cutting_drive_motor_3_defect': 'Cutting drive motor 3 defect',
    'cutting_height_blocked': 'Cutting height blocked',
    'cutting_height_problem': 'Cutting height problem',
    'cutting_height_problem_curr': 'Cutting height problem, curr',
    'cutting_height_problem_dir': 'Cutting height problem, dir',
    'cutting_height_problem_drive': 'Cutting height problem, drive',
    'cutting_motor_problem': 'Cutting motor problem',
    'cutting_stopped_slope_too_steep': 'Cutting stopped - slope too steep',
    'cutting_system_blocked': 'Cutting system blocked',
    'cutting_system_imbalance_warning': 'Cutting system imbalance    Warning',
    'cutting_system_major_imbalance': 'Cutting system major imbalance',
    'destination_not_reachable': 'Destination not reachable',
    'difficult_finding_home': 'Difficult finding home',
    'docking_sensor_defect': 'Docking sensor defect',
    'electronic_problem': 'Electronic problem',
    'empty_battery': 'Empty battery',
    'folding_cutting_deck_sensor_defect': 'Folding cutting deck sensor defect',
    'folding_sensor_activated': 'Folding sensor activated',
    'geofence_problem': 'Geofence problem',
    'gps_navigation_problem': 'GPS navigation problem',
    'guide_1_not_found': 'Guide 1 not found',
    'guide_2_not_found': 'Guide 2 not found',
    'guide_3_not_found': 'Guide 3 not found',
    'guide_calibration_accomplished': 'Guide calibration accomplished',
    'guide_calibration_failed': 'Guide calibration failed',
    'high_charging_power_loss': 'High charging power loss',
    'high_internal_power_loss': 'High internal power loss',
    'high_internal_temperature': 'High internal temperature',
    'internal_voltage_error': 'Internal voltage error',
    'invalid_battery_combination_invalid_combination_of_different_battery_types': 'Invalid battery combination - Invalid combination of different battery types.',
    'invalid_sub_device_combination': 'Invalid sub-device combination',
    'invalid_system_configuration': 'Invalid system configuration',
    'left_brush_motor_overloaded': 'Left brush motor overloaded',
    'lift_sensor_defect': 'Lift Sensor defect',
    'lifted': 'Lifted',
    'limited_cutting_height_range': 'Limited cutting height range',
    'loop_sensor_defect': 'Loop sensor defect',
    'loop_sensor_problem_front': 'Loop sensor problem, front',
    'loop_sensor_problem_left': 'Loop sensor problem, left',
    'loop_sensor_problem_rear': 'Loop sensor problem, rear',
    'loop_sensor_problem_right': 'Loop sensor problem, right',
    'low_battery': 'Low battery',
    'memory_circuit_problem': 'Memory circuit problem',
    'mower_lifted': 'Mower lifted',
    'mower_tilted': 'Mower tilted',
    'no_accurate_position_from_satellites': 'No accurate position from satellites',
    'no_confirmed_position': 'No confirmed position',
    'no_drive': 'No drive',
    'no_loop_signal': 'No loop signal',
    'no_power_in_charging_station': 'No power in charging station',
    'no_response_from_charger': 'No response from charger',
    'outside_working_area': 'Outside working area',
    'poor_signal_quality': 'Poor signal quality',
    'reference_station_communication_problem': 'Reference station communication problem',
    'right_brush_motor_overloaded': 'Right brush motor overloaded',
    'safety_function_faulty': 'Safety function faulty',
    'settings_restored': 'Settings restored',
    'sim_card_locked': 'SIM card locked',
    'sim_card_not_found': 'SIM card not found',
    'sim_card_requires_pin': 'SIM card requires PIN',
    'slipped_mower_has_slipped_situation_not_solved_with_moving_pattern': 'Slipped - Mower has Slipped.Situation not solved with moving pattern',
    'slope_too_steep': 'Slope too steep',
    'sms_could_not_be_sent': 'SMS could not be sent',
    'stop_button_problem': 'STOP button problem',
    'stuck_in_charging_station': 'Stuck in charging station',
    'switch_cord_problem': 'Switch cord problem',
    'temporary_battery_problem': 'Temporary battery problem',
    'tilt_sensor_problem': 'Tilt sensor problem',
    'too_high_discharge_current': 'Too high discharge current',
    'too_high_internal_current': 'Too high internal current',
    'trapped': 'Trapped',
    'ultrasonic_problem': 'Ultrasonic problem',
    'ultrasonic_sensor_1_defect': 'Ultrasonic Sensor 1 defect',
    'ultrasonic_sensor_2_defect': 'Ultrasonic Sensor 2 defect',
    'ultrasonic_sensor_3_defect': 'Ultrasonic Sensor 3 defect',
    'ultrasonic_sensor_4_defect': 'Ultrasonic Sensor 4 defect',
    'unexpected_cutting_height_adj': 'Unexpected cutting height adj',
    'unexpected_error': 'Unexpected error',
    'upside_down': 'Upside down',
    'weak_gps_signal': 'Weak GPS signal',
    'wheel_drive_problem_left': 'Wheel drive problem, left',
    'wheel_drive_problem_rear_left': 'Wheel drive problem, rear left',
    'wheel_drive_problem_rear_right': 'Wheel drive problem, rear right',
    'wheel_drive_problem_right': 'Wheel drive problem, right',
    'wheel_motor_blocked_left': 'Wheel motor blocked, left',
    'wheel_motor_blocked_rear_left': 'Wheel motor blocked, rear left',
    'wheel_motor_blocked_rear_right': 'Wheel motor blocked, rear right',
    'wheel_motor_blocked_right': 'Wheel motor blocked, right',
    'wheel_motor_overloaded_left': 'Wheel motor overloaded, left',
    'wheel_motor_overloaded_rear_left': 'Wheel motor overloaded, rear left',
    'wheel_motor_overloaded_rear_right': 'Wheel motor overloaded, rear right',
    'wheel_motor_overloaded_right': 'Wheel motor overloaded, right',
    'work_area_not_valid': 'Work area not valid',
    'wrong_loop_signal': 'Wrong loop signal',
    'wrong_pin_code': 'Wrong PIN code',
    'zone_generator_problem': 'Zone generator problem',
  })
# ---
# name: test_error_keys_snapshot
  list([
    'alarm_mower_in_motion',
    'alarm_mower_lifted',
    'alarm_mower_stopped',
    'alarm_mower_switched_off',
    'alarm_mower_tilted',
    'alarm_outside_geofence',
    'angular_sensor_problem',
    'battery_problem',
    'battery_problem',
    'battery_restriction_due_to_ambient_temperature',
    'can_error',
    'charging_current_too_high',
    'charging_station_blocked',
    'charging_system_problem',
    'charging_system_problem',
    'collision_sensor_defect',
    'collision_sensor_error',
    'collision_sensor_problem_front',
    'collision_sensor_problem_rear',
    'com_board_not_available',
    'communication_circuit_board_sw_must_be_updated',
    'complex_working_area',
    'connection_changed',
    'connection_not_changed',
    'connectivity_problem',
    'connectivity_problem',
    'connectivity_problem',
    'connectivity_problem',
    'connectivity_problem',
    'connectivity_problem',
    'connectivity_settings_restored',
    'cutting_drive_motor_1_defect',
    'cutting_drive_motor_2_defect',
    'cutting_drive_motor_3_defect',
    'cutting_height_blocked',
    'cutting_height_problem',
    'cutting_height_problem_curr',
    'cutting_height_problem_dir',
    'cutting_height_problem_drive',
    'cutting_motor_problem',
    'cutting_stopped_slope_too_steep',
    'cutting_system_blocked',
    'cutting_system_blocked',
    'cutting_system_imbalance_warning',
    'cutting_system_major_imbalance',
    'destination_not_reachable',
    'difficult_finding_home',
    'docking_sensor_defect',
    'electronic_problem',
    'empty_battery',
    'folding_cutting_deck_sensor_defect',
    'folding_sensor_activated',
    'geofence_problem',
    'geofence_problem',
    'gps_navigation_problem',
    'guide_1_not_found',
    'guide_2_not_found',
    'guide_3_not_found',
    'guide_calibration_accomplished',
    'guide_calibration_failed',
    'high_charging_power_loss',
    'high_internal_power_loss',
    'high_internal_temperature',
    'internal_voltage_error',
    'invalid_battery_combination_invalid_combination_of_different_battery_types',
    'invalid_sub_device_combination',
    'invalid_system_configuration',
    'left_brush_motor_overloaded',
    'lift_sensor_defect',
    'lifted',
    'limited_cutting_height_range',
    'limited_cutting_height_range',
    'loop_sensor_defect',
    'loop_sensor_problem_front',
    'loop_sensor_problem_left',
    'loop_sensor_problem_rear',
    'loop_sensor_problem_right',
    'low_battery',
    'memory_circuit_problem',
    'mower_lifted',
    'mower_tilted',
    'no_accurate_position_from_satellites',
    'no_confirmed_position',
    'no_drive',
    'no_loop_signal',
    'no_power_in_charging_station',
    'no_response_from_charger',
    'outside_working_area',
    'poor_signal_quality',
    'reference_station_communication_problem',
    'right_brush_motor_overloaded',
    'safety_function_faulty',
    'settings_restored',
    'sim_card_locked',
    'sim_card_locked',
    'sim_card_locked',
    'sim_card_locked',
    'sim_card_not_found',
    'sim_card_requires_pin',
    'slipped_mower_has_slipped_situation_not_solved_with_moving_pattern',
    'slope_too_steep',
    'sms_could_not_be_sent',
    'stop_button_problem',
    'stuck_in_charging_station',
    'switch_cord_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'temporary_battery_problem',
    'tilt_sensor_problem',
    'too_high_discharge_current',
    'too_high_internal_current',
    'trapped',
    'ultrasonic_problem',
    'ultrasonic_sensor_1_defect',
    'ultrasonic_sensor_2_defect',
    'ultrasonic_sensor_3_defect',
    'ultrasonic_sensor_4_defect',
    'unexpected_cutting_height_adj',
    'unexpected_error',
    'upside_down',
    'weak_gps_signal',
    'wheel_drive_problem_left',
    'wheel_drive_problem_rear_left',
    'wheel_drive_problem_rear_right',
    'wheel_drive_problem_right',
    'wheel_motor_blocked_left',
    'wheel_motor_blocked_rear_left',
    'wheel_motor_blocked_rear_right',
    'wheel_motor_blocked_right',
    'wheel_motor_overloaded_left',
    'wheel_motor_overloaded_rear_left',
    'wheel_motor_overloaded_rear_right',
    'wheel_motor_overloaded_right',
    'work_area_not_valid',
    'wrong_loop_signal',
    'wrong_pin_code',
    'zone_generator_problem',
  ])
# ---
